File: miscellaneous/search/20230405/binary-search.js
function binarySearch(values, target, start, end) {
    start = start ?? 0;
    end = end ?? values.length - 1;

    const mid = start + Math.floor((end - start) / 2);

    if (target < values[mid]) return binarySearch(values, target, 0, mid);
    else if (target > values[mid]) return binarySearch(values, target, mid, end);
    else if (values[mid] === target) return mid;
}

module.exports = binarySearch;

File: miscellaneous/search/20230405/binary_search.py
import math
from typing import Tuple


def binarySearch(nums: Tuple[int], target: int, start: int = None, end: int = None) -> int:
    start = 0 if start is None else start = start
    end = len(nums) - 1 if end is None else end = end

    if end - start < 0:
        return -1

    mid = math.floor((start - end) / 2) + start

    if target < nums[mid]:
        return binarySearch(nums, target, start, mid)

    if target > nums[mid]:
        return binarySearch(nums, target, mid)

    if target == nums[mid]:
        return mid

    return -1



